### PR TITLE
Time Series Insights - GA api-version 2020-05-15

### PR DESCRIPTION
**Pull request by riserrad**
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [X] I have reviewed the [documentation](https://github.com/Azure/adx-documentation-pr/wiki/Overall-basic-flow) for the workflow.
- [X] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [X] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [ ] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

### CODE REVIEW

**Comment by ArcturusZhang** [comment]
Hi @riserrad please let me know when this is ready to review

**Comment by riserrad** [comment]
> Hi @riserrad please let me know when this is ready to review

Sure, @ArcturusZhang , it should be ready as soon as I get another pair of eyes from my manager on it.

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (sandshadow) [inline specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json]: Should be "expand" now, instead of "$expand", right?

**Review by sandshadow — CHANGES_REQUESTED** [review]

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (majastrz) [inline specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json]: >version [](start = 5, length = 7)

Is there a previous API version that was approved for Time Series Insights that I can compare this with?

Comment (sandshadow) [reply specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json]: This one should be almost identical to the [2018-08-15-preview](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/preview/2018-08-15-preview/timeseriesinsights.json) public preview api-version

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (majastrz) [inline specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json]: >HTTP 400 (Bad Request):  [](start = 28, length = 24)

The `default` response is a catch-all for all kinds of errors (4xx and 5xx), so this is somewhat misleading.

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (majastrz) [inline specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json]: >404 [](start = 11, length = 3)

What's the reason for explicitly declaring a 404 response? Most swaggers rely purely on the `default` response.

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (majastrz) [reply specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json]: Thanks that helps a lot.

---
In reply to: [400576000](https://github.com/Azure/azure-rest-api-specs/pull/8791#discussion_r400576000) [](ancestors = 400576000)

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (majastrz) [reply specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json]: This applies to all occurrences of this pattern in this swagger.

---
In reply to: [400578474](https://github.com/Azure/azure-rest-api-specs/pull/8791#discussion_r400578474) [](ancestors = 400578474)

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (majastrz) [inline specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json]: >The environment was successfully deleted. [](start = 28, length = 41)

204 is supposed to be returned when the resource doesn't exist at the time of the request. Can you update the description?

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (majastrz) [inline specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json]: >get [](start = 7, length = 3)

You should declare paging via the x-ms-pageable extension ([https://github.com/Azure/autorest/blob/master/docs/extensions/readme.md#x-ms-pageable](https://github.com/Azure/autorest/blob/master/docs/extensions/readme.md#x-ms-pageable)).

It's ok if your backend doesn't support paging yet (it's effectively always returning a single page), but making this change makes it easy to add paging to the backend in the future without breaking any client or complicating the implementation.

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (majastrz) [inline specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json]: >get [](start = 7, length = 3)

Same paging comment as above.

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (majastrz) [inline specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json]: >eventSources [](start = 142, length = 12)

Are you going to need paging on eventSources?

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (majastrz) [inline specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json]: >referenceDataSets [](start = 142, length = 17)

Do you need paging?

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (majastrz) [inline specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json]: >accessPolicies [](start = 142, length = 14)

Paging?

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (majastrz) [inline specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json]: >creationTime [](start = 9, length = 12)

Should be marked readOnly

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (majastrz) [inline specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json]: >keyName [](start = 9, length = 7)

This is not a secret, right?

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (majastrz) [reply specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json]: Applies to all payloads that declare this property.

---
In reply to: [400581394](https://github.com/Azure/azure-rest-api-specs/pull/8791#discussion_r400581394) [](ancestors = 400581394)

**Comment by majastrz** [comment]
I added a bunch of comments. Please take a look.

**Comment by riserrad** [comment]
Thanks @majastrz . I will look into your comments.

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (riserrad) [reply specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json:123]: HTTP 404 will happen specifically when either the subscription or resource group, passed as part of the URL - not the request body, is not found.

@sandshadow what do you think of this suggestion, though? If we go for it, then the `default` description becomes misleading instead.

I'd stick with having a separate 404.

Comment (majastrz) [reply specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json:123]: 404 is not the only error that is possible. If you were to go down this way, you would have to document every possible intentional and unintentional HTTP status code. That list would include but is not limited to 400, 403, 404, 429, 500, 502, 503, etc. It is not scalable to have every Swagger include this information when it's common across all RPs. As such the convention is to declare only the `default` error.

On top of that declaring non-`default` error responses has side effects on the SDKs that get generated from this swagger and makes them behave in unexpected ways.

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (riserrad) [reply specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json:548]: I don't think people will have a number of Event Sources on a single Environment that would activate paging.

Would it hurt if we leave the x-ms-pageable for Event Sources as well? I'm thinking of standardizing.

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (riserrad) [reply specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json:2060]: No, `sharedAccessKey` - a bit below - is. Any recommendations there?

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (riserrad) [reply specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json:795]: Same as Event Source here.

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (riserrad) [reply specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json:1042]: Here, paging definitely makes sense.

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (riserrad) [reply specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json:1196]: Does the line below with `"readOnly": true` satisfy this? Or should we do it differently?

Comment (majastrz) [reply specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json:1208]: Yeah that works. I must have missed it before.

---
In reply to: [401791458](https://github.com/Azure/azure-rest-api-specs/pull/8791#discussion_r401791458) [](ancestors = 401791458)

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (majastrz) [inline specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json:2214]: >sharedAccessKey [](start = 9, length = 15)

Can you add `"x-ms-secret": true` to this property to mark it as a secret?

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (majastrz) [inline specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json:2229]: >sharedAccessKey [](start = 9, length = 15)

Can you add `"x-ms-secret": true` to this property to mark it as a secret?

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (majastrz) [inline specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json:2092]: >sharedAccessKey [](start = 9, length = 15)

Can you add `"x-ms-secret": true` to this property to mark it as a secret?

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (majastrz) [reply specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json:2072]: Yeah, you should add `"x-ms-secret": true` to any property that is a secret.

---
In reply to: [401787589](https://github.com/Azure/azure-rest-api-specs/pull/8791#discussion_r401787589) [](ancestors = 401787589)

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (majastrz) [reply specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json]: ARM tries to follow OData (a tiny subset, really), so this should actually be `$expand`.

---
In reply to: [398956944](https://github.com/Azure/azure-rest-api-specs/pull/8791#discussion_r398956944) [](ancestors = 398956944)

**Comment by majastrz** [comment]
Previous comments appear to be resolved. I added a few more. Please take a look.

File: specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json
Comment (sandshadow) [reply specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json]: Agreed - and the parameter used to be $expand just to meet OData standards as you say. However, ARM broke that underneath us some time last spring - calls  that used the "$expand" started failing with an error saying something like "$" is a reserved character. Since analysis showed this query parameter was only really used by our Ibiza extension, we switched to "expand" to fix the issue.
I had intended to raise this issue with ARM, but it fell off my plate.

Comment (majastrz) [reply specification/timeseriesinsights/resource-manager/Microsoft.TimeSeriesInsights/stable/2020-05-15/timeseriesinsights.json]: I don't believe that ARM was the cause of that bug because there were millions of successful requests with $expand in the last hour alone.

---
In reply to: [401840532](https://github.com/Azure/azure-rest-api-specs/pull/8791#discussion_r401840532) [](ancestors = 401840532)

**Comment by riserrad** [comment]
@majastrz all feedbacks were addressed and all automated checks are now passing.

**Comment by majastrz** [comment]
I see all as resolved except one (it's buried somewhere in the GitHub UI, so it probably got missed). 

Regarding, the $expand->expand change, ARM should not be the cause of that bug because there are millions of successful requests with $expand in the last hour alone. On top of that my team owns two different RPs that use $expand and they appear to be working fine without any issues. 

Thoughts?

**Comment by riserrad** [comment]
> I see all as resolved except one (it's buried somewhere in the GitHub UI, so it probably got missed).
> 
> Regarding, the $expand->expand change, ARM should not be the cause of that bug because there are millions of successful requests with $expand in the last hour alone. On top of that my team owns two different RPs that use $expand and they appear to be working fine without any issues.
> 
> Thoughts?

Right, I understand your points and we can definitely work on putting it back as `$expand`. But my question, for now, would be: is it a deal-breaker for this PR?

The reason for asking is that it would require some work on our backend to adjust it - adjusting, testing, deploying, etc - and, if it's not a deal-breaker, I'd rather stick with the `expand` at least for this version.

**Review by majastrz — APPROVED** [review]
Signed off from ARM side with a comment. Please look into why `$expand` is not working for you but is working for millions of requests every hour and switch it back when you figure it out.

**Comment by riserrad** [comment]
Please don't check this PR yet. I will solve this last issue so we can have a 100% approved change.

The update will be there in a few minutes.

**Comment by riserrad** [comment]
Please, wait for an update from us to check this PR in. I will share updates soon.

**Comment by majastrz** [comment]
@riserrad I added the "do not merge" label. When you're ready, just let @ArcturusZhang know when it's ok to merge it.

**Comment by ArcturusZhang** [comment]
Hi @riserrad any update on this?

**Comment by riserrad** [comment]
> Hi @riserrad any update on this?

Hi @ArcturusZhang , not yet. We will probably have this PR carried over the next week.

Is there any problem with that? Do you have any recommendations?

**Comment by ArcturusZhang** [comment]
Hi @riserrad do we have any update on this PR? It has been a long time....

**Comment by riserrad** [comment]
> Hi @riserrad do we have any update on this PR? It has been a long time....

Hi @ArcturusZhang , you're right. And as we're still take a little while to open this api-version, I'd say we should abandon this PR and I'll open a new one when we're good to go.

Thanks for the follow up.

**Pull request closed without merging**
